docs/2024/puzzles/day15.md

```diff
@@ -2,10 +2,402 @@ import Solver from "../../../../../website/src/components/Solver.js"
 
 # Day 15: Warehouse Woes
 
+by [@shardulc](https://github.com/shardulc)
+
 ## Puzzle description
 
 https://adventofcode.com/2024/day/15
 
+## Solution
+
+### Summary
+
+When building a grid-based simulation like for this puzzle, my preferred approach is to use an immutable 2D array representation for the grid and make an updated copy at every step. Then, to run the simulation is simply to compute a *fold* of the moves over the (grid) state. I value the ease (for me) of programming and reasoning about the resulting program over the run-time costs of allocation, garbage collection, etc.
+
+(With a mutable array, we’d have to be careful not to, say, overwrite an old box before it has been moved to its new position, and I don’t trust myself to reason that carefully. We could alternatively use a collection of objects with their positions instead of an array, which would make more sense for sparse grids, as finding adjacent objects would require traversing the entire collection.)
+
+Some preliminaries: enums to represent grid cells and move directions…
+
+```scala
+enum WarehouseCell:
+  case Empty
+  case Wall
+  case Box
+  case LeftBox
+  case RightBox
+  case Robot
+
+enum Direction:
+  case Up
+  case Down
+  case Left
+  case Right
+```
+
+…and a class to represent the state. In addition to the grid, we choose to (redundantly) track the position of the robot, since there’s only one and we might want its position often and don’t want to traverse the grid to find it every time.
+
+```scala
+class Warehouse(
+    val cells: ArraySeq[ArraySeq[WarehouseCell]],
+    val robotRow: Int,
+    val robotCol: Int)
+```
+
+*NB:* We use `ArraySeq`s instead of `Array`s to guarantee that we don’t accidentally mutate them in our program while still getting constant-time random access.
+
+### Part 1
+
+We start with the parser. First, we’ll need a way to create `Direction` and `WarehouseCell` objects from their `Char` representations. We could either write a simple pattern matcher:
+
+```scala
+object Direction:
+  def ofChar(c: Char) =
+    c match
+      case '^' => Up
+      case 'v' => Down
+      case '>' => Right
+      case '<' => Left
+```
+
+Or we could make the enum parametric, and use the parameter to define a `Map` (which we can automatically use as a `Char => WarehouseCell` function due to its `apply` method):
+
+```scala
+enum WarehouseCell(val repr: Char):
+  case Empty extends WarehouseCell('.')
+  case Wall extends WarehouseCell('#')
+  case Box extends WarehouseCell('O')
+  case LeftBox extends WarehouseCell('[')
+  case RightBox extends WarehouseCell(']')
```

Comment:
Remove? The start of the part 2 section mentions adjusting this model to account for the big boxes.